ROX-10084: Add dirty tag filter for operator installation

kurlov opened the pull request:
## Description

[Ticket description](https://issues.redhat.com/browse/ROX-10084)

Dirty tag helps to spot if there are some changes since last commit. Such tags are not pushed to the image registry. Running `olm-operator-install` with dirty tag stuck operator in `ImagePullBackOff` loop because there is no such image in the registry. I think it makes sense to fail fast dirty tag olm installation and give developer a hint why this is not going to work.



## Checklist
- [x] Investigated and inspected CI test results
~~- [ ] Unit test and regression tests added~~
~~- [ ] Evaluated and added CHANGELOG entry if required~~
~~- [ ] Determined and documented upgrade steps~~
~~- [ ] Documented user facing changes (create PR based on [stackrox/openshift-docs](https://github.com/stackrox/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

No need, it is just a quality of life improvement

## Testing Performed

Run locally with dirty tag against remote cluster and it failed immediately

kurlov commented:
`gke-ui-e2e-tests ` - flaky test

porridge commented on operator/hack/common.sh:
```diff
@@ -65,6 +65,15 @@ function retry() {
+function check_version_tag() {
+  local -r version_tag="$1"
+
+    if [[ $version_tag == *-dirty ]]; then
+      log "Cannot install from *-dirty image tag. Dirty tag images are not supposed to be pushed to the registry"
+      return 1
+    fi
```
Please use consistent indentation
```suggestion
  if [[ $version_tag == *-dirty ]]; then
    log "Cannot install from *-dirty image tag. Dirty tag images are not supposed to be pushed to the registry."
    return 1
  fi
```

porridge reviewed (commented):
I'm not sure I agree with the "there is no reason" point in the PR description. I think there might be a use case for both ways - for example committing one's operator-only changes causes the tag for the stacrox main image to change and cause a rebuilt, which is useless but takes time.

Are you're saying trying to deploy the `-dirty` tag does not actually work? If so, do you know what the reason is?

Of course, there is also a good reason why using a `-dirty` tag is not a great idea (if two developers start off the same `master` commit and don't commit their changes, they will end up with the same tag and clash with confusing symptoms.

So I think it might be more healthy long-term to disallow this, but the PR description does not seem to sufficiently describe the reasons.

kurlov commented:
> Are you're saying trying to deploy the -dirty tag does not actually work? If so, do you know what the reason is?

Yes. it will not work. At least olm installation part. The problem is that `olm-operator-install` command relies on the existence of the image in the registry with the current tag. Usually, `-dirty` tags are not pushed to the registry, Thus such command will fail later on the image pull stage.

I guess somebody might push dirty tag manually from the local machine and then run `olm-operator-install` command. Then such check from the PR will be blocking. Probably, I should extend `check_version_tag` function and check if image is presented in the registry. And in case of missing image fail immediately. I think it would be better. @porridge what do you think?

I have updated the PR description.

porridge commented:
> > Are you're saying trying to deploy the -dirty tag does not actually work? If so, do you know what the reason is?
> 
> Yes. it will not work. At least olm installation part. The problem is that `olm-operator-install` command relies on the existence of the image in the registry with the current tag.

Well, this depends on what cluster you're deploying to, right?
What you said is true for a remote (i.e. in the cloud) cluster. For example a docker-desktop-based k8s cluster it will happily pick up whatever images are present in your local docker daemon store, I think.

> Usually, `-dirty` tags are not pushed to the registry

Not sure what you mean here. Are you saying that our scripts somehow treat `-dirty` specially?
Or merely that CI will never see a `-dirty` tag by definition (because it always runs on a clean checkout - well even that is subtle, because if one forgets to commit autogenerated changes then CI will likely dirty its working tree, but the checks in place _should_ prevent it from proceeding before it gets to pushing images).

> I guess somebody might push dirty tag manually from the local machine and then run `olm-operator-install` command. Then such check from the PR will be blocking.

Not sure what you mean here. As written above, I don't think CI will ever get to see a `-dirty` tag, if that's what you mean by "check from the PR"...

> Probably, I should extend `check_version_tag` function and check if image is presented in the registry. And in case of missing image fail immediately. I think it would be better. @porridge what do you think?

Yeah, I think that would be better if not for the case of the local cluster :-/
How about a check that can be overridden with an environment variable?

SimonBaeumer commented:
@porridge 

> Or merely that CI will never see a -dirty tag by definition
Yeah, CI only sometimes pushed dirty tags. I never deployed dirty tag deployed by CI though.

> How about a check that can be overridden with an environment variable?

+1, and guiding users to set the env var from the error message like "... to deploy dirty tag set ENV_VAR=..."

kurlov commented:
> Well, this depends on what cluster you're deploying to, right?
What you said is true for a remote (i.e. in the cloud) cluster. For example a docker-desktop-based k8s cluster it will happily pick up whatever images are present in your local docker daemon store, I think.

Good catch. I think checking if it's a local cluster or not might be ambiguous.

> Not sure what you mean here. As written above, I don't think CI will ever get to see a -dirty tag, if that's what you mean by "check from the PR"...

Yes, you are right. 

I really like the idea of introducing specific variable. I added it and extended makefile accordingly. Also, I added check for docker image existence in the registry. If image exists then it doesn't matter is it dirty or not. In that case we can ignore filter and proceed with the installation.

porridge commented on operator/Makefile:
```diff
@@ -321,19 +326,28 @@ undeploy: check-ci-setup kustomize ## Undeploy operator image from the K8s clust
+.PHONY: deploy-dirty-tag-via-olm
+deploy-dirty-tag-via-olm: ## Deploy operator dirty tagged image to the cluster using OLM.
+# This command ignores filter for dirty tagged images.
+	ALLOW_DIRTY_TAG_INSTALLATION ?=true KUTTL=$(KUTTL) ./hack/olm-operator-install.sh $(TEST_NAMESPACE) $(VERSION) true
```
```suggestion
	KUTTL=$(KUTTL) ./hack/olm-operator-install.sh $(TEST_NAMESPACE) $(VERSION) true
```

porridge commented on operator/Makefile:
```diff
@@ -321,19 +326,28 @@ undeploy: check-ci-setup kustomize ## Undeploy operator image from the K8s clust
+
+.PHONY: deploy-dirty-tag-via-olm
+deploy-dirty-tag-via-olm: ## Deploy operator dirty tagged image to the cluster using OLM.
+# This command ignores filter for dirty tagged images.
```
```suggestion
# This target ignores filter for dirty tagged images.
```

porridge commented on operator/Makefile:
```diff
@@ -321,19 +326,28 @@ undeploy: check-ci-setup kustomize ## Undeploy operator image from the K8s clust
+	KUTTL=$(KUTTL) ./hack/olm-operator-upgrade.sh $(TEST_NAMESPACE) $(VERSION) $(ALLOW_DIRTY_TAG_INSTALLATION)
+
+.PHONY: upgrade-dirty-tag-via-olm
+upgrade-via-olm: kuttl
```
```suggestion
upgrade-dirty-tag-via-olm: kuttl
```

porridge commented on operator/hack/common.sh:
```diff
@@ -65,6 +65,26 @@ function retry() {
+
+
+  if [[ $version_tag == *-dirty ]] && [[ $allow_dirty_tag = false ]]; then
+    log "Cannot install from *-dirty image tag. Please, use 'deploy-dirty-tag-via-olm' command instead if you need to install dirty tagged image"
```
This message seems to be specific to one makefile target, but can be emitted in other contexts too.
Perhaps mentioning the variable would make more sense, but then someone might be invoking the scripts directly rather than via `make`, where the variable does not apply.

kurlov replied:
You are right. I also added additional log just to explicitly note if it's dirty tagged image

porridge reviewed (changes requested):
I'm not sure we need the extra Makefile targets, maybe using the variable would be sufficient. But I'm not going to oppose it if you think they are useful.
Also a bunch of requests inline.

SimonBaeumer commented on operator/hack/common.sh:
```diff
@@ -65,6 +65,26 @@ function retry() {
+  fi
+
+
+  if [[ $version_tag == *-dirty ]] && [[ $allow_dirty_tag = false ]]; then
```
Please add quotes to variables. 
Also should the if contain `==` equals?
```suggestion
  if [[ "$version_tag" == *-dirty ]] && [[ "$allow_dirty_tag" == false ]]; then
```

kurlov commented on operator/hack/olm-operator-install.sh:
```diff
@@ -7,20 +7,23 @@ source "$(dirname "$0")/common.sh"
     ;;
   *)
-    echo "Usage: $0 <operator_ns> <index-version> [<install-version>]" >&2
+    echo "Usage: $0 <operator_ns> <index-version> <allow-dirty-tag> [<install-version>]" >&2
```
Good idea. I have made it as an optional flag. No need to set true or false just flag itself is enough.

porridge commented on operator/hack/common.sh:
```diff
@@ -65,6 +65,28 @@ function retry() {
+    log "Cannot find $IMAGE_TAG_BASE:$version_tag image tag in the registry."
+  fi
+
+  if [[ "$version_tag" == *-dirty ]]; then
```
Note that (unlike with `[ ]`) with `[[ ]]` the quotes around arguments are actually optional even if the variable contains whitespace. I think shellcheck will even complain about them.
cc @SimonBaeumer who I think suggested this

SimonBaeumer replied:
Shellcheck did  not complain on my machine. I always add quotes around variables to avoid evaluating variables as shell commands.

kurlov replied:
Shellcheck does not complain for me neither. Although there is such rule [SC2053](https://github.com/koalaman/shellcheck/wiki/SC2053)  but it's not applicable here

porridge replied:
Looks like I remembered this wrong, don't worry about this.

porridge commented on operator/hack/common.sh:
```diff
@@ -65,6 +65,28 @@ function retry() {
+  local -r version_tag="$1"
+  local -r allow_dirty_tag="$2"
+
+  echo $IMAGE_TAG_BASE
```
Leftover debug code?

porridge commented on operator/hack/common.sh:
```diff
@@ -65,6 +65,28 @@ function retry() {
+  local -r allow_dirty_tag="$2"
+
+  echo $IMAGE_TAG_BASE
+  if docker manifest inspect "$IMAGE_TAG_BASE:$version_tag" ; then
```
What's the purpose of this? In case of remote cluster deployment the existence of an image locally does not mean much, it might be missing in the image repo, leading to image pull errors and failing deployment... 🤔

kurlov replied:
It checks if image is pushed to the registry. `IMAGE_TAG_BASE` equals to `docker.io/stackrox/stackrox-operator` by default. Thus `docker manifest inspect` checks image in the registry, not locally. 
If image is there then it doesn't matter is it a dirty tagged or not so developer will not be bothered to use different command for installing an operator. It's quite fast command and basically just a tiny improvement.

kurlov replied:
You are right, I did not think about such scenario that image is built locally and haven't pushed and installation is triggered. Well, It's was pretty ambiguous check. I removed it

SimonBaeumer reviewed (approved):
LGTM from my side. Please address @porridge comments and 🚢 .

porridge reviewed (approved):
You might also want to mention these new targets in the README if you think that will be useful.
LGTM otherwise!
